fancyerror.go

```diff
@@ -0,0 +1,256 @@
+// It's reasonably large because:
+// a) HTML often has large embedded scripts which we want
+// to skip and
+// b) it should be an relatively unusual case so the size
+// shouldn't harm.
+//
+// It's defined as a variable so that it can be redefined in tests.
+var maxErrorBodySize = 200 * 1024
+
+// isJSONMediaType reports whether the content type of the given header implies
+// that the content is JSON.
+func isJSONMediaType(header http.Header) bool {
+	contentType := header.Get("Content-Type")
+	mediaType, _, _ := mime.ParseMediaType(contentType)
+	return mediaType == "application/json"
+}
+
+// Error implements error.Error by trying to produce a decent
+// error message derived from the body content.
+func (e *fancyDecodeError) Error() string {
+	mediaType, _, err := mime.ParseMediaType(e.contentType)
+	if err != nil {
+		// Even if there's no media type, we want to see something useful.
+		mediaType = fmt.Sprintf("%q", e.contentType)
+	}
+
+	// TODO use charset.NewReader to convert from non-utf8 content?
+	switch mediaType {
+	case "text/html":
+		text, err := htmlToText(bytes.NewReader(e.body))
+		if err != nil {
+			// Note: it seems that this can never actually
+			// happen - the only way that the HTML parser
+			// can fail is if there's a read error and we've
+			// removed that possibility by using
+			// noErrorReader above.
+			return fmt.Sprintf("unexpected (and invalid) content text/html; want application/json; content: %q", sizeLimit(e.body))
+		}
+		if len(text) == 0 {
+			return fmt.Sprintf(`unexpected content type text/html; want application/json; content: %q`, sizeLimit(e.body))
+		}
+		return fmt.Sprintf(`unexpected content type text/html; want application/json; content: %s`, sizeLimit(text))
+	case "text/plain":
+		return fmt.Sprintf(`unexpected content type text/plain; want application/json; content: %s`, sizeLimit(sanitizeText(string(e.body), true)))
+	default:
+		return fmt.Sprintf(`unexpected content type %s; want application/json; content: %q`, mediaType, sizeLimit(e.body))
+	}
+}
+
+// noErrorReader wraps a reader, turning any errors into io.EOF
+// so that we can extract some content even if we get an io error.
+type noErrorReader struct {
+	r io.Reader
+}
+
+func (r noErrorReader) Read(buf []byte) (int, error) {
+	n, err := r.r.Read(buf)
+	if err != nil {
+		err = io.EOF
+	}
+	return n, err
+}
+
+func sizeLimit(data []byte) []byte {
+	const max = 1024
+	if len(data) < max {
+		return data
+	}
+	return append(data[0:max], fmt.Sprintf(" ... [%d bytes omitted]", len(data)-max)...)
+}
+
+// htmlToText attempts to return some relevant textual content
+// from the HTML content in the given reader, formatted
+// as a single line.
+func htmlToText(r io.Reader) ([]byte, error) {
+	n, err := html.Parse(r)
+	if err != nil {
+		return nil, err
+	}
+	var buf bytes.Buffer
+	htmlNodeToText(&buf, n)
+	return buf.Bytes(), nil
+}
+
+func htmlNodeToText(w *bytes.Buffer, n *html.Node) {
```

Comment:
Can I suggest putting a doc comment here to say exactly how this tries to extract text. I can see infinite possibilities for how to do this conversion, but I don't want to bikeshed an error path.